### ISSUE TITLE
"Save to default" feature

### ISSUE THREAD

**Issue by OverLordGoldDragon** [spyder-ide/spyder#13736]
Often I wish to run / edit a code snippet in a fresh file, and not lose it - but don't seek to save it in any particular directory, especially not the project / previous file's directory. It'd help to have a settable "default" directory, just like a default file title (Untitled), to which one could save with a dedicated command - thus not having to manually search for one (and bypassing any selection prompts).

**Comment by steff456** [comment]
Hi @OverLordGoldDragon,

We need to discuss the importance of this feature request with the rest of the team. I'll let you know what we decide.

**Comment by steff456** [comment]
Hi @OverLordGoldDragon,

Can you please provide us a concrete example of this functionality? Like the user cases you'll like to add

Thanks!

**Comment by OverLordGoldDragon** [comment]
@steff456 Sure; [demo](https://puu.sh/GqysB/0fbe01b2dd.mp4):

  1. Pasted code snippet into a fresh tab
  2. Edited it, and let's pretend I did stuff with it
  3. Would like to save it, but am prompted to select a save directory
  4. Spyder automatically suggests the folder of the previous file (`main.py`), which is undesired

So instead, there could be an extra option to save to a default directory - and Spyder would increment the name appropriately if needed (e.g. `untitled4` -> `untitled5`):

<img src="https://user-images.githubusercontent.com/16495490/92513844-0088da00-f222-11ea-98a2-d3e78dc0bd47.png" height="80">

**Comment by ccordoba12** [comment]
This looks very confusing and I don't think we'll add something like that.

I understand your idea of having scratchpad kind of files, but we'd need to come up with a better user experience for that concept.

**Comment by OverLordGoldDragon** [comment]
@ccordoba12 The idea is to have a one shortcut/click save of a file to a user-defined "default" directory; doesn't seem more confusing than configuring the working directory or startup script. 

One _could_ simplify to saving it to `.spyder-py3/defaultsave`, just like `/autosave`, with maybe some notification of where the save went to, and "You can configure this directory in Preferences", with the "Don't show this again" option - so the user's informed of how the feature works. ... actually this is less simple but it sure ought to be less confusing.

**Comment by isabela-pf** [comment]
Hey @OverLordGoldDragon, thanks for taking the time to post this issue and giving insight into a part of how you work in Spyder! I want to make sure I’m understanding the problem well and that we look into a solution that will resolve the root of the problem, so I have a few clarifying questions.
1. Could you tell me more about when you want to save a code snippet in a fresh file without keeping it near the project it originated from? What kind of code stands out? Why save it separately? What do you do with that code once you have it separate from the original project?
2. Why have a directory automatically created by Spyder over making a `default` (or similar) directory created by a user that they can choose from the Save as? It sounds like because it takes more than a single click or keyboard shortcut, but are there other reasons too?
3. Is sharing content across files ever a goal in this process? Does the code you want to quickly save end up in multiple different documents, or does it tend to just stay in it's own file to not be used again?

Thanks again!

**Comment by OverLordGoldDragon** [comment]
@isabela-pf Thanks for taking interest.

  1. It's to not have to open a fresh new Spyder, with its own project (or lack of). Example: I'm unit-testing a package, and someone shares an unrelated code I awaited. I need to edit the code several times and re-run it with a fresh namespace; also rather not lose it by accident. This requires saving.

  2. The user most likely won't know how the feature works, and won't have said directory configured. So, Spyder could (1) make its own `default` and simply notify the user that that's where the file went to, and that it can be changed via Preferences, or (2) prompt the user to select such a directory. (2) involves an extra step, and may be less beginner-friendly, but is otherwise fine.

  3. Standalone. Though, it could draw from files in the same default directory if the "working directory" is set to the file's.

**Comment by isabela-pf** [comment]
I agree this could be useful and a simple interaction. There are a have a few UI changes that would need to be implemented to support it. 

First, I agree it should appear In the File menu as previously mentioned. I also agree that grouping it in the menu with other save options is the right choice.

Second, it makes sense to provide preferences for the user to configure save location and file name. I’ve also shown what sound like they should be the default options: save to some kind of unique, Spyder designated directory (I made up one) and give the default name Untitled. It should look like this: 

<img width="1058" alt="Preferences" src="https://user-images.githubusercontent.com/50221806/94646168-85f02f00-02a2-11eb-9e18-d918fae5b36c.png">

Third, I don’t know what shortcut would make sense for this, but having it at least able to be assigned a shortcut manually sounds important to minimize the actions needed for a quick save.

Talking with other Spyder team members brought up the question of how to handle these files if they are open when Spyder is closed. I think they can function like all other files and reopen the next time Spyder is opened.

Feedback is always welcome, especially since I might be missing the nuances of this feature.

**Comment by OverLordGoldDragon** [comment]
@isabela-pf Excellent, all looks good and better; `quicksave`'s more succinct. Unsure what's meant by "open when Spyder is closed", but I'd go with "just like any other save".

Didn't expect this to gain traction - glad it's being seriously considered. Lately I've had `untitled`s piling up in projects, and fun times renumbering them all when moving to a common directory.

Unsure if PyCharm has this; could be a thing to include in those "Spyder vs PyCharm" articles.

**Comment by oscargus** [comment]
I just stumbled upon this and I realize that I would also use this feature every now and then. Typically, if I want to debug something (using the debugger, so just pasting in the terminal is not feasible) and run a minimal example (a couple of lines or so) that are only relevant while debugging.

I have quite a few of these files hanging around in my home directory, Desktop, and so on. Of course it is easy to avoid (by deleting them afterwards etc), but some sort of scratch-pad would indeed be useful. I can even imagine storing the file in a temporary directory, although I will probably at some stage regret that and I realize that it may be a bad idea from a general user perspective...

Here it is not so much about projects and environments though, more that I am lazy...

**Comment by BeeMaan** [comment]
I really enjoyed using Spyder but when my PC unexpectedly crashed and lost my work I sought out alternatives. The way some other IDEs implement data loss prevention might offer ideas. PyCharm effectively requires a script to be named (ie saved) before it can be run (generally any script worth saving will have been run at least once to see how it runs, although its amazing how much you can end up having typed out in even just 30 mins and then forget to make a save). VSCode will keep persistent and reopen whatever open tabs there were even after a fatal shutdown subject to having been preserved within the time window of its user-set autosave. This is probably the ideal realtime auto/quick-save of whatever state the IDE is in at any given time

**Comment by ccordoba12** [comment]
Hey @BeeMaan, thanks for reporting. We'll try to improve how we handle untitled files in Spyder 6.

**Comment by OverLordGoldDragon** [comment]
FWIW I've since discovered

```python
%edit
```

which does its own thing (see docs), and which I've extended and can invoke as follows:

```python
epath = gen_epath()
%edit {epath}
```

with `gen_epath` being available from the startup script (`Preferences > IPython Console > Startup > Run a file`). With Spyder's "up arrow" feature, I just type `ep`, arrow up, then enter, and done - opens a file in an expected directory.

<details><summary><b>gen_epath</b></summary>

```python
import os, random


def gen_epath():
    _dir = r"C:\Desktop\code\dev"
    while True:
        # Generate four random digits
        random_digits = "".join(str(random.randint(0, 9)) for _ in range(4))
        fname = os.path.join(_dir, 'dev_' + str(random_digits) + '.py')

        # Check if the file already exists
        if not os.path.exists(fname):
            return fname
```
</details>
